src/standard.ts

```diff
@@ -1,202 +1,196 @@
-import { equalsBytes, hexToBytes } from 'ethereum-cryptography/utils';
-import { Bytes, compareBytes, hex } from './bytes';
-import { getProof, isValidMerkleTree, makeMerkleTree, processProof, renderMerkleTree, MultiProof, getMultiProof, processMultiProof } from './core';
-import { MerkleTreeOptions, defaultOptions } from './options';
-import { checkBounds } from './utils/check-bounds';
-import { throwError } from './utils/throw-error';
-import { standardLeafHash } from './utils/standard-leaf-hash';
-
-interface StandardMerkleTreeData<T extends any[]> {
-  format: 'standard-v1';
-  tree: string[];
-  values: {
-    value: T;
-    treeIndex: number;
-  }[];
-  leafEncoding: string[];
-}
+import { BytesLike, HexString, toHex, isBytes, isBytesLike } from "./bytes";
 
-export class StandardMerkleTree<T extends any[]> {
-  private readonly hashLookup: { [hash: string]: number };
+import { MultiProof, processProof, processMultiProof } from "./core";
 
-  private constructor(
-    private readonly tree: Bytes[],
-    private readonly values: { value: T, treeIndex: number }[],
-    private readonly leafEncoding: string[],
-  ) {
-    this.hashLookup =
-      Object.fromEntries(values.map(({ value }, valueIndex) => [
-        hex(standardLeafHash(value, leafEncoding)),
-        valueIndex,
-      ]));
-  }
+import {
+  SimpleMerkleTree,
+  SimpleMerkleTreeData,
+  SimpleMerkleTreeValue,
+} from "./simple";
 
-  static of<T extends any[]>(values: T[], leafEncoding: string[], options: MerkleTreeOptions = {}) {
-    const sortLeaves = options.sortLeaves ?? defaultOptions.sortLeaves;
+import { defaultAbiCoder } from "@ethersproject/abi";
+import { keccak256 } from "@ethersproject/keccak256";
 
-    const hashedValues = values.map((value, valueIndex) => ({ value, valueIndex, hash: standardLeafHash(value, leafEncoding) }));
+import { StandardMerkleTreeOptions } from "./options";
+import { throwError } from "./utils/throw-error";
 
-    if (sortLeaves) {
-      hashedValues.sort((a, b) => compareBytes(a.hash, b.hash));
-    }
+export type StandardMerkleTreeValue = SimpleMerkleTreeValue;
 
-    const tree = makeMerkleTree(hashedValues.map(v => v.hash));
+export type StandardMerkleTreeData = SimpleMerkleTreeData & {
+  leafEncoding?: string[];
+};
 
-    const indexedValues = values.map(value => ({ value, treeIndex: 0 }));
-    for (const [leafIndex, { valueIndex }] of hashedValues.entries()) {
-      indexedValues[valueIndex]!.treeIndex = tree.length - leafIndex - 1;
-    }
+export function standardLeafHash<T extends BytesLike[] = BytesLike[]>(
```

Comment:
Is `T extends BytesLike[]` ok? The values could be numbers, arrays, structures, basically anything supported by the AbiEncoder.